// @braze/i18n-scripts/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseTypescriptParametersFromTranslation = exports.unzipTranslations = exports.getChangedJSONItems = exports.getTranslationChanges = exports.syncToTranslationRepo = exports.syncToDevRepo = exports.maxDepthTranslations = exports.deepMerge = exports.generateMergedTranslations = exports.sortMergedTranslations = exports.generateIndexDTsFromTranslations = exports.saveConfig = exports.loadConfig = exports.addNewLocale = exports.mergeAllTranslations = exports.updateTranslations = void 0;
// index.ts for node context
const config_node_1 = require("./config-node");
Object.defineProperty(exports, "loadConfig", { enumerable: true, get: function () { return config_node_1.loadConfig; } });
Object.defineProperty(exports, "saveConfig", { enumerable: true, get: function () { return config_node_1.saveConfig; } });
const updateTranslations_1 = require("./utils/updateTranslations");
Object.defineProperty(exports, "updateTranslations", { enumerable: true, get: function () { return updateTranslations_1.updateTranslations; } });
const mergeAllTranslations_1 = require("./utils/mergeAllTranslations");
Object.defineProperty(exports, "mergeAllTranslations", { enumerable: true, get: function () { return mergeAllTranslations_1.mergeAllTranslations; } });
const generateIndexDTsFromTranslations_1 = require("./utils/generateIndexDTsFromTranslations");
Object.defineProperty(exports, "generateIndexDTsFromTranslations", { enumerable: true, get: function () { return generateIndexDTsFromTranslations_1.generateIndexDTsFromTranslations; } });
const generateMergedTranslations_1 = require("./utils/generateMergedTranslations");
Object.defineProperty(exports, "generateMergedTranslations", { enumerable: true, get: function () { return generateMergedTranslations_1.generateMergedTranslations; } });
const sortMergedTranslations_1 = require("./utils/sortMergedTranslations");
Object.defineProperty(exports, "sortMergedTranslations", { enumerable: true, get: function () { return sortMergedTranslations_1.sortMergedTranslations; } });
const parseTypescriptParametersFromTranslation_1 = require("./utils/parseTypescriptParametersFromTranslation");
Object.defineProperty(exports, "parseTypescriptParametersFromTranslation", { enumerable: true, get: function () { return parseTypescriptParametersFromTranslation_1.parseTypescriptParametersFromTranslation; } });
const addNewLocale_1 = require("./utils/addNewLocale");
Object.defineProperty(exports, "addNewLocale", { enumerable: true, get: function () { return addNewLocale_1.addNewLocale; } });
const deepMerge_1 = require("./utils/deepMerge");
Object.defineProperty(exports, "deepMerge", { enumerable: true, get: function () { return deepMerge_1.deepMerge; } });
Object.defineProperty(exports, "maxDepthTranslations", { enumerable: true, get: function () { return deepMerge_1.maxDepthTranslations; } });
const sync_to_dev_1 = __importDefault(require("./sync-translation-repos/sync-to-dev"));
exports.syncToDevRepo = sync_to_dev_1.default;
const sync_to_translation_1 = __importDefault(require("./sync-translation-repos/sync-to-translation"));
exports.syncToTranslationRepo = sync_to_translation_1.default;
const get_translation_changes_1 = __importStar(require("./translation-changes/get-translation-changes"));
exports.getTranslationChanges = get_translation_changes_1.default;
Object.defineProperty(exports, "getChangedJSONItems", { enumerable: true, get: function () { return get_translation_changes_1.getChangedJSONItems; } });
const unzip_translations_1 = __importDefault(require("./translation-changes/unzip-translations"));
exports.unzipTranslations = unzip_translations_1.default;
